test/src/module/protocol/protocolTest.c

```diff
@@ -535,193 +535,92 @@ testRun(void)
-                TEST_RESULT_VOID(ioWriteStrLine(write, STRDEF("{\"cmd\":\"r-c\"}")), "write complex request");
-                TEST_RESULT_VOID(ioWriteFlush(write), "flush complex request");
-                TEST_RESULT_STR_Z(ioReadLine(read), "{\"out\":false}", "complex request result");
-                TEST_RESULT_STR_Z(ioReadLine(read), ".LINEOFTEXT", "complex request result");
-                TEST_RESULT_STR_Z(ioReadLine(read), ".", "complex request result");
-
-                // Exit
-                TEST_RESULT_VOID(ioWriteStrLine(write, STRDEF("{\"cmd\":\"exit\"}")), "write exit");
-                TEST_RESULT_VOID(ioWriteFlush(write), "flush exit");
-
-                // Retry errors until success
-                TEST_RESULT_VOID(ioWriteStrLine(write, STRDEF("{\"cmd\":\"ezero\"}")), "write error-until-0");
-                TEST_RESULT_VOID(ioWriteFlush(write), "flush error-until-0");
-                TEST_RESULT_STR_Z(ioReadLine(read), "{\"out\":true}", "error-until-0 result");
-
-                // Exit
-                TEST_RESULT_VOID(ioWriteStrLine(write, STRDEF("{\"cmd\":\"exit\"}")), "write exit");
-                TEST_RESULT_VOID(ioWriteFlush(write), "flush exit");
-            }
-            HARNESS_FORK_CHILD_END();
+                    pckReadStrP(protocolClientExecute(client, protocolCommandNew(TEST_PROTOCOL_COMMAND_SIMPLE), true)), "output",
+                    "execute");
 
-            HARNESS_FORK_PARENT_BEGIN()
-            {
-                IoRead *read = ioFdReadNew(STRDEF("server read"), HARNESS_FORK_PARENT_READ_PROCESS(0), 2000);
-                ioReadOpen(read);
-                IoWrite *write = ioFdWriteNew(STRDEF("server write"), HARNESS_FORK_PARENT_WRITE_PROCESS(0), 2000);
-                ioWriteOpen(write);
+                // -----------------------------------------------------------------------------------------------------------------
+                TEST_TITLE("complex command");
 
-                // Send greeting
-                ProtocolServer *server = NULL;
+                ProtocolCommand *command = NULL;
+                TEST_ASSIGN(command, protocolCommandNew(TEST_PROTOCOL_COMMAND_COMPLEX), "command");
+                TEST_RESULT_VOID(pckWriteU32P(protocolCommandParam(command), 87), "param");
+                TEST_RESULT_VOID(pckWriteStrP(protocolCommandParam(command), STRDEF("data")), "param");
 
-                MEM_CONTEXT_TEMP_BEGIN()
-                {
-                    TEST_ASSIGN(
-                        server,
-                        protocolServerMove(
-                            protocolServerNew(STRDEF("test server"), STRDEF("test"), read, write), memContextPrior()),
-                        "create server");
-                    TEST_RESULT_VOID(protocolServerMove(NULL, memContextPrior()), "move null server");
-                }
-                MEM_CONTEXT_TEMP_END();
+                TEST_RESULT_VOID(protocolClientCommandPut(client, command), "command put");
```

Comment:
So here we "Put command to the server" but then the next statement is a  protocolClientDataGet(client) and I thought that would be "Get data put by the server" but instead you indicate "sync" - what is that? I'm having a hard time understanding what is going on in this test.

Reply:
This is the server letting the client know it is ready to receive data. In some cases it makes no sense for the client to send data until the server has verified that, e.g. the file to write can be opened. Added a comment at e66caa8431.